lib/services/app.go

```diff
@@ -158,3 +164,109 @@ func UnmarshalAppServer(data []byte, opts ...MarshalOption) (types.AppServer, er
 	}
 	return nil, trace.BadParameter("unsupported app server resource version %q", h.Version)
 }
+
+// NewApplicationFromKubeService creates application resources from kubernetes service.
+// It transforms service fields and annotations into appropriate Teleport app fields.
+// Service labels are copied to app labels.
+func NewApplicationFromKubeService(service corev1.Service, clusterName, protocol string, port corev1.ServicePort) (types.Application, error) {
+	appURI := buildAppURI(protocol, getServiceFQDN(service), port.Port)
+
+	rewriteConfig, err := getAppRewriteConfig(service.GetAnnotations())
+	if err != nil {
+		return nil, trace.Wrap(err, "could not get app rewrite config for the service")
+	}
+
+	appNameAnnotation := service.GetAnnotations()[types.DiscoveryAppNameLabel]
+	appName, err := getAppName(service.GetName(), service.GetNamespace(), clusterName, port.Name, appNameAnnotation)
+	if err != nil {
+		return nil, trace.Wrap(err, "could not create app name for the service")
+	}
+
+	labels, err := getAppLabels(service.GetLabels(), clusterName)
+	if err != nil {
+		return nil, trace.Wrap(err, "could not get labels for the service")
+	}
+
+	app, err := types.NewAppV3(types.Metadata{
+		Name:        appName,
+		Description: fmt.Sprintf("Discovered application in Kubernetes cluster %q", clusterName),
+		Labels:      labels,
+	}, types.AppSpecV3{
+		URI:     appURI,
+		Rewrite: rewriteConfig,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err, "could not create an app from Kubernetes service")
+	}
+
+	return app, nil
+}
+
+func getServiceFQDN(s corev1.Service) string {
+	host := s.GetName()
+	if s.Spec.Type == corev1.ServiceTypeExternalName {
```

Comment:
Can't we always use the internal service DNS? 
We let KubeDNS handle the translation of the internal -> external endpoint.